src/Storages/Kafka/StorageKafka2.cpp

```diff
@@ -0,0 +1,1490 @@
+                virtual_columns[7]->insert(headers_values);
+                if (put_error_to_stream)
+                {
+                    if (exception_message)
+                    {
+                        virtual_columns[8]->insert(consumer.currentPayload());
+                        virtual_columns[9]->insert(*exception_message);
+                    }
+                    else
+                    {
+                        virtual_columns[8]->insertDefault();
+                        virtual_columns[9]->insertDefault();
+                    }
+                }
+            }
+
+            total_rows = total_rows + new_rows;
+            batch_info.last_offset = consumer.currentOffset();
+        }
+        else if (consumer.isStalled())
+        {
+            ++failed_poll_attempts;
+        }
+        else
+        {
+            // We came here in case of tombstone (or sometimes zero-length) messages, and it is not something abnormal
+            // TODO: it seems like in case of put_error_to_stream=true we may need to process those differently
+            // currently we just skip them with note in logs.
+            LOG_DEBUG(
+                log,
+                "Parsing of message (topic: {}, partition: {}, offset: {}) return no rows.",
+                consumer.currentTopic(),
+                consumer.currentPartition(),
+                consumer.currentOffset());
+        }
+
+        if (!consumer.hasMorePolledMessages()
+            && (total_rows >= getMaxBlockSize() || !check_time_limit() || failed_poll_attempts >= MAX_FAILED_POLL_ATTEMPTS
+                || consumer.needsOffsetUpdate()))
+        {
+            LOG_TRACE(
+                log,
+                "Stopped collecting message for current batch. There are {} failed polled attempts, {} total rows and consumer needs "
+                "offset update is {}",
+                failed_poll_attempts,
+                total_rows,
+                consumer.needsOffsetUpdate());
+            break;
+        }
+    }
+
+    if (total_rows == 0)
+        return {};
+
+    /// MATERIALIZED columns can be added here, but I think
+    // they are not needed here:
+    // and it's misleading to use them here,
+    // as columns 'materialized' that way stays 'ephemeral'
+    // i.e. will not be stored anythere
+    // If needed any extra columns can be added using DEFAULT they can be added at MV level if needed.
+
+    auto result_block = non_virtual_header.cloneWithColumns(executor.getResultColumns());
+    auto virtual_block = virtual_header.cloneWithColumns(std::move(virtual_columns));
+
+    for (const auto & column : virtual_block.getColumnsWithTypeAndName())
+        result_block.insert(column);
+
+    batch_info.blocks.emplace_back(std::move(result_block));
+    return batch_info;
+}
+
+void StorageKafka2::threadFunc(size_t idx)
+{
+    chassert(idx < tasks.size());
+    auto task = tasks[idx];
+    std::optional<StallReason> maybe_stall_reason;
+    try
```

Comment:
Reschedule here if `!is_active`?

Reply:
I am not sure what do you mean here. `is_active` is set to `false` in `partialShutdown`, which deactivates the streaming tasks, thus these task should be stopped soon after `is_active` is set to `false`. Because of this, I didn't think we should care about `is_active` here. If I miss something, please let me know what do I miss.